Make <progress> use Page's preferredRenderingUpdateInterval()

annevk opened the pull request:
#### ae044324e12b0f44593288217af847856a8405b1
<pre>
Make &lt;progress&gt; use Page&apos;s preferredRenderingUpdateInterval()
<a href="https://bugs.webkit.org/show_bug.cgi?id=265277">https://bugs.webkit.org/show_bug.cgi?id=265277</a>

Reviewed by Darin Adler.

This ensures that we take various other considerations into account
when animating and not just call repaint() at a fixed repeat interval.

This has the side effect of making the animation on macOS run at 60FPS
and on visionOS at 90FPS, under normal circumstances.

For now we don&apos;t change where the repeat interval is obtained from, as
it&apos;s not immediately clear to me that a separate RenderTheme bool
method as to whether &lt;progress&gt; animates is preferable to the existing
setup.

* Source/WebCore/rendering/RenderProgress.cpp:
(WebCore::RenderProgress::animationTimerFired):
(WebCore::RenderProgress::updateAnimationState):
* Source/WebCore/rendering/RenderProgress.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::animationRepeatIntervalForProgressBar const): Deleted.
(WebCore::RenderTheme::animationDurationForProgressBar const): Deleted.
(WebCore::RenderTheme::progressBarRectForBounds const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::animationRepeatIntervalForProgressBar const):
(WebCore::RenderTheme::animationDurationForProgressBar const):
(WebCore::RenderTheme::progressBarRectForBounds const):
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::animationRepeatIntervalForProgressBar const):
(WebCore::RenderThemeAdwaita::animationDurationForProgressBar const):
* Source/WebCore/rendering/RenderThemeAdwaita.h:
* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::animationRepeatIntervalForProgressBar const):
* Source/WebCore/rendering/RenderThemeIOS.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::animationRepeatIntervalForProgressBar const): Deleted.
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::animationRepeatIntervalForProgressBar const): Deleted.

Canonical link: <a href="https://commits.webkit.org/271431@main">https://commits.webkit.org/271431@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e341b72e73927812259cad83e035146126e6fa0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28367 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30894 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28864 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9106 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4381 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24402 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5052 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5151 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25403 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31580 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25975 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25840 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31453 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5115 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3291 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29209 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6714 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6800 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5570 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5635 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderProgress.cpp:
```diff
@@ -81,27 +81,30 @@ bool RenderProgress::isDeterminate() const
 {
-    // FIXME: Progress bar animation should be performed as part of the rendering update
-    // lifecycle, to match the display's refresh rate.
+    // FIXME: ideally obtaining the repeat interval from Page is not RenderTheme-specific, but it
```
WebKit coding style says capital letter: "Ideally".

darinadler commented on Source/WebCore/rendering/RenderProgress.cpp:
```diff
@@ -81,27 +81,30 @@ bool RenderProgress::isDeterminate() const
+    // FIXME: ideally obtaining the repeat interval from Page is not RenderTheme-specific, but it
+    // currently is as it also determines whether we animate at all.
+    auto refreshInterval = theme().animationRepeatIntervalForProgressBar(*this);
+    if (!(refreshInterval > 0_s))
```
Unclear why this has to check at all; `updateAnimationState` checks this and sets `m_animating` appropriately. No explanation in the change log or comments why this additional check is needed. Maybe because `updateAnimationState` is not called in some cases? Seems to me that if the interval is now 0, we would like `m_animating` to now be `false`.

I’m not sure I understand why an early return without `repaint` is correct for a refresh interval of 0 when a timer happens to have been set. If `m_animating` is false we do the repaint and don’t set a new timer. Seems risky to skip the `repaint` if the timer had been set before, although maybe OK.

annevk replied:
The reason I added this is because refreshInterval could in theory become 0 due to the element or page disappearing or some such. And I don't think we want to have this code run in essentially a loop at that point.

Maybe you are correct though and `updateAnimationState()` would have been invoked in those cases.

@pxlcoder can you weigh in perhaps?

darinadler replied:
If this is defensive coding for a case that we don’t know how to reproduce, then it might be safer to have it check after the `repaint` call, to guarantee we don’t leave something behind? Or maybe it’s better to check before like this does to avoid trying to do any work inside `repaint` that might lead to a crash. Or maybe `repaint` is guaranteed not to do anything in that case.

Beyond trying to write code that makes few assumptions, the best way to deal with a worry about the element or page going away is to think about how to construct a test. I hate to risk having such a consideration slow down an otherwise straightforward and important change, but I can’t help seeing the looseness of the idiom here.

annevk commented on Source/WebCore/rendering/RenderTheme.h:
```diff
@@ -215,11 +215,9 @@ class RenderTheme {
-    virtual Seconds animationDurationForProgressBar(const RenderProgress&) const;
-    virtual IntRect progressBarRectForBounds(const RenderProgress&, const IntRect&) const;
+    virtual Seconds animationRepeatIntervalForProgressBar(const RenderProgress&) const { return 0_s; }
+    virtual Seconds animationDurationForProgressBar(const RenderProgress&) const { return 0_s; }
```
Nit: remove the RenderProgress argument. No need to pass it around until a theme actually needs it.

annevk committed:
Make <progress> use Page's preferredRenderingUpdateInterval()
https://bugs.webkit.org/show_bug.cgi?id=265277

Reviewed by Darin Adler.

This ensures that we take various other considerations into account
when animating and not just call repaint() at a fixed repeat interval.

This has the side effect of making the animation on macOS run at 60FPS
and on visionOS at 90FPS, under normal circumstances.

For now we don't change where the repeat interval is obtained from, as
it's not immediately clear to me that a separate RenderTheme bool
method as to whether <progress> animates is preferable to the existing
setup.

* Source/WebCore/rendering/RenderProgress.cpp:
(WebCore::RenderProgress::animationTimerFired):
(WebCore::RenderProgress::updateAnimationState):
* Source/WebCore/rendering/RenderProgress.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::animationRepeatIntervalForProgressBar const): Deleted.
(WebCore::RenderTheme::animationDurationForProgressBar const): Deleted.
(WebCore::RenderTheme::progressBarRectForBounds const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::animationRepeatIntervalForProgressBar const):
(WebCore::RenderTheme::animationDurationForProgressBar const):
(WebCore::RenderTheme::progressBarRectForBounds const):
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::animationRepeatIntervalForProgressBar const):
(WebCore::RenderThemeAdwaita::animationDurationForProgressBar const):
* Source/WebCore/rendering/RenderThemeAdwaita.h:
* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::animationRepeatIntervalForProgressBar const):
* Source/WebCore/rendering/RenderThemeIOS.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::animationRepeatIntervalForProgressBar const): Deleted.
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::animationRepeatIntervalForProgressBar const): Deleted.

Canonical link: https://commits.webkit.org/271431@main
